### PR TITLE
Add cmdlets to change cli Settings of ESG

### LINKED ISSUE #431: No change ESG or DLR admin password from PowerNSX
Whilst it is possible to change credentials from GUI or API, PowerNSX does not expose the <password> element of <cliSettings> meaning passwords cannot be changed.

### PR DESCRIPTION
**Pull request by alagoutte**
Add 2 new cmdlets for get and set cliSettings of ESG (or DLR)

Fix #431

### CODE REVIEW

**Comment by alagoutte** [comment]
There is not yet testsuite but i will be follow soon

**Comment by alagoutte** [comment]
Hi,

Now with testsuite (and also add a testsuite for Enable/DisableSSH)

Can you review ?

**Comment by alagoutte** [comment]
jenkins test this please

**Comment by nmbradford** [comment]
Hi Alexis - sorry for the large delay responding - only getting a chance to actually catch up on PowerNSX backlog now.  Ive taken a quick look at the merge - and a few things occur.

Firstly - I think this should be a part of set-nsxedge rather than separate cmdlets.  I recently did some policy work and as an example, check out how set-nsxsecuritypolicy works - it can either accept updated XML, OR explicitly update specific properties.  best of both worlds.  I think this is the model that all set-xxx cmdlets should use now.  To that end, set-nsxedge already does the first (Accepts updated XML) and I think this an all related base (appliance, cli etc) edge config should be handled here too.  So your code should be pretty much copy/paste to update set-nsxedge... and your tests should be easy to move too...

**Comment by nmbradford** [comment]
Ill add a few other minor comments in review too...

File: module/PowerNSX.psm1 [hunk 1]
```diff
@@ -13969,6 +14001,174 @@ function Disable-NsxEdgeSsh {
 
 }
 
+function Get-NsxcliSettings {
```
Comment (nmbradford) [inline module/PowerNSX.psm1]: No need for this cmdlet if making part of set-nsxedge...  If it is to be a seperate cmdlet though, the name should have the word edge in it!

File: module/PowerNSX.psm1 [hunk 2]
```diff
@@ -13969,6 +14001,174 @@ function Disable-NsxEdgeSsh {
...
+
+    param (
+
+        [Parameter (Mandatory=$true,ValueFromPipeline=$true,Position=1)]
+            [ValidateScript({ ValidateEdge $_ })]
+            [System.Xml.XmlElement]$Edge
+    )
+
+    begin {
+
```
Comment (nmbradford) [inline module/PowerNSX.psm1]: kill whitespace

File: module/PowerNSX.psm1 [hunk 3]
```diff
@@ -13969,6 +14001,174 @@ function Disable-NsxEdgeSsh {
...
+    .DESCRIPTION
+    An NSX Edge Service Gateway provides all NSX Edge services such as firewall,
+    NAT, DHCP, VPN, load balancing, and high availability. Each NSX Edge virtual
+    appliance can have a total of ten uplink and internal network interfaces and
+    up to 200 subinterfaces.  Multiple external IP addresses can be configured
+    for load balancer, site‐to‐site VPN, and NAT services.
+
+    The Set-NsxcliSettings cmdlet configure the cli Settings of the ESG
+    it is mandatory to specified the password...
```
Comment (nmbradford) [inline module/PowerNSX.psm1]: Why?  any other edge xml update doesnt require this?  are you saying no cli settings can be updated unless credential is specified?  I havent tested this yet, but it sounds wrong.

File: module/PowerNSX.psm1 [hunk 4]
```diff
@@ -13969,6 +14001,174 @@ function Disable-NsxEdgeSsh {
...
+    [Diagnostics.CodeAnalysis.SuppressMessageAttribute("PSAvoidDefaultValueSwitchParameter","")] # Cant remove without breaking backward compatibility
+    param (
+
+        [Parameter (Mandatory=$true,ValueFromPipeline=$true,Position=1)]
+            [ValidateScript({ ValidateCliSettings $_ })]
+            [System.Xml.XmlElement]$cliSettings,
+        [Parameter (Mandatory=$false)]
+            [ValidateNotNullorEmpty()]
+            [String]$userName,
+        [Parameter (Mandatory=$true)]
```
Comment (nmbradford) [inline module/PowerNSX.psm1]: as per above.

File: module/PowerNSX.psm1 [hunk 5]
```diff
@@ -13969,6 +14001,174 @@ function Disable-NsxEdgeSsh {
...
+            [ValidateRange(1,99999)]
+            [int]$passwordExpiry,
+        [Parameter (Mandatory=$false)]
+            [ValidateNotNullorEmpty()]
+            [string]$sshLoginBannerText,
+        [Parameter (Mandatory=$False)]
+            #PowerNSX Connection object
+            [ValidateNotNullOrEmpty()]
+            [PSCustomObject]$Connection=$defaultNSXConnection
+
```
Comment (nmbradford) [inline module/PowerNSX.psm1]: kill whitespace

File: module/PowerNSX.psm1 [hunk 6]
```diff
@@ -13969,6 +14001,174 @@ function Disable-NsxEdgeSsh {
...
+    )
+
+    begin {
+
```
Comment (nmbradford) [inline module/PowerNSX.psm1]: kill whitespace

File: module/PowerNSX.psm1 [hunk 7]
```diff
@@ -13969,6 +14001,174 @@ function Disable-NsxEdgeSsh {
...
+        #Create private xml element
+        $_cliSettings = $cliSettings.CloneNode($true)
+
+        #Store the edgeId and remove it from the XML as we need to post it...
+        $edgeId = $_cliSettings.edgeId
+        $_cliSettings.RemoveChild( $((Invoke-XPathQuery -QueryMethod SelectSingleNode -Node $_cliSettings -Query 'descendant::edgeId')) ) | out-null
+
+        #Using PSBoundParamters.ContainsKey lets us know if the user called us with a given parameter.
+        #If the user did not specify a given parameter, we dont want to modify from the existing value.
+
```
Comment (nmbradford) [inline module/PowerNSX.psm1]: kill whitespace

File: module/PowerNSX.psm1 [hunk 8]
```diff
@@ -13969,6 +14001,174 @@ function Disable-NsxEdgeSsh {
...
+        if ( $PsBoundParameters.ContainsKey('userName') ) {
+            $_cliSettings.username = $userName
+        }
+
+        #You need ALWAYS to specified the password...
```
Comment (nmbradford) [inline module/PowerNSX.psm1]: as per above

File: module/PowerNSX.psm1 [hunk 9]
```diff
@@ -13969,6 +14001,174 @@ function Disable-NsxEdgeSsh {
...
+        Add-XmlElement -xmlRoot $_cliSettings -xmlElementName "password" -xmlElementText $password
+
+        if ( $PsBoundParameters.ContainsKey('remoteAccess') ) {
+            if ( $remoteAccess ) {
```
Comment (nmbradford) [inline module/PowerNSX.psm1]: just use $remoteAccess.ToString().ToLower() instead of the if/else

File: module/PowerNSX.psm1 [hunk 10]
```diff
@@ -13969,6 +14001,174 @@ function Disable-NsxEdgeSsh {
...
+
+        #Store the edgeId and remove it from the XML as we need to post it...
+        $edgeId = $_cliSettings.edgeId
+        $_cliSettings.RemoveChild( $((Invoke-XPathQuery -QueryMethod SelectSingleNode -Node $_cliSettings -Query 'descendant::edgeId')) ) | out-null
+
+        #Using PSBoundParamters.ContainsKey lets us know if the user called us with a given parameter.
+        #If the user did not specify a given parameter, we dont want to modify from the existing value.
+
+        if ( $PsBoundParameters.ContainsKey('userName') ) {
+            $_cliSettings.username = $userName
```
Comment (nmbradford) [inline module/PowerNSX.psm1]: I also get nervous where I see assumption that a specific XML node exists.  Please get into habit of testing for and creating the element if it doesnt exist - i get bitten by this too often.

File: tests/integration/04.Edge.Tests.ps1 [hunk 11]
```diff
@@ -660,6 +660,67 @@ Describe "Edge" {
         }
     }
 
+    Context "SSH" {
```
Comment (nmbradford) [inline tests/integration/04.Edge.Tests.ps1]: Great that you are adding tests, but why are you including ssh tests here?  should be separate PR...

File: tests/integration/04.Edge.Tests.ps1 [hunk 12]
```diff
@@ -660,6 +660,67 @@ Describe "Edge" {
...
+    }
+
+    Context "CliSettings" {
+
+        it "Can disable SSH" {
+            $edge = Get-NsxEdge $name
+            #When deploy pstester ESG, the SSH is enabled
+            $edge.cliSettings.remoteAccess | should be "true"
+            #Need Always to set a password to change cliSettings
+            Get-NsxEdge $name | Get-NsxcliSettings | Set-NsxCliSettings -password Vmware1!Vmware1! -remoteAccess:$false
```
Comment (nmbradford) [inline tests/integration/04.Edge.Tests.ps1]: you havent tested get-nsxclisettings yet - you need prior test that tests this (and you should use it two lines above too!) - this will probably be a moot point if you move these functions to set-nsxedge though...

File: tests/integration/04.Edge.Tests.ps1 [hunk 13]
```diff
@@ -660,6 +660,67 @@ Describe "Edge" {
...
+        it "Can enable SSH" {
+            Get-NsxEdge $name | Get-NsxcliSettings | Set-NsxCliSettings -password Vmware1!Vmware1! -remoteAccess:$true
+            $edge = Get-NsxEdge $name
+            $edge.cliSettings.remoteAccess | should be "true"
+        }
+
+        it "Change (SSH) username" {
+            $edge = Get-NsxEdge $name
+            #By default it is admin
+            $edge.cliSettings.userName | should be "admin"
```
Comment (nmbradford) [inline tests/integration/04.Edge.Tests.ps1]: if we dont rely on it, then no point in testing for it...

File: tests/integration/04.Edge.Tests.ps1 [hunk 14]
```diff
@@ -660,6 +660,67 @@ Describe "Edge" {
...
+            Get-NsxEdge $name | Get-NsxcliSettings | Set-NsxCliSettings -password Vmware1!Vmware1! -userName powernsx
+            $edge = Get-NsxEdge $name
+            $edge.cliSettings.userName | should be "powernsx"
+        }
+
+        it "Change Password Expiry" {
+            $edge = Get-NsxEdge $name
+            #By default it is 99999
+            $edge.cliSettings.passwordExpiry | should be "99999"
```
Comment (nmbradford) [inline tests/integration/04.Edge.Tests.ps1]: likewise

File: tests/integration/04.Edge.Tests.ps1 [hunk 15]
```diff
@@ -660,6 +660,67 @@ Describe "Edge" {
...
+            Get-NsxEdge $name | Get-NsxcliSettings | Set-NsxCliSettings -password Vmware1!Vmware1! -passwordExpiry 42
```
Comment (nmbradford) [inline tests/integration/04.Edge.Tests.ps1]: haha! :D

File: tests/integration/04.Edge.Tests.ps1 (hunk 11 above)
Comment (alagoutte) [reply tests/integration/04.Edge.Tests.ps1]: Ok
it is already a separate commit... (i see this issue when write my testsuite...)

File: tests/integration/04.Edge.Tests.ps1 (hunk 12 above)
Comment (alagoutte) [reply tests/integration/04.Edge.Tests.ps1]: do you have a example ?

File: module/PowerNSX.psm1 (hunk 3 above)
Comment (alagoutte) [reply module/PowerNSX.psm1]: Try and you will see that... (may be crazy when see that...)

**Comment by alagoutte** [comment]
not sure it is a good idea to include on set-nsxedge it will be a big cmdlets with a lot of parameter..
and when there is some direct API, it is better to use ? (and more faster ? like send all nsx-edge info)

or for rename to *-NsxEdgecliSettings !

File: module/PowerNSX.psm1 (hunk 3 above)
Comment (nmbradford) [reply module/PowerNSX.psm1]: i tested updating the banner text with the clisettings api - and yes it rejects it without credentials, _however_, i also tried it by updating the full edge api - and it works fine.  Another reason for including in set-nsxedge :D

File: tests/integration/04.Edge.Tests.ps1 (hunk 11 above)
Comment (nmbradford) [reply tests/integration/04.Edge.Tests.ps1]: its because you aren't branching from master.  you should use master to track upstream, branch from it for every isolated PR you intend to raise.  For this case, dont worry - _as long as the PR tests pass when we execute the full suite against it_

File: tests/integration/04.Edge.Tests.ps1 (hunk 12 above)
Comment (nmbradford) [reply tests/integration/04.Edge.Tests.ps1]: all the other tests that test retrieval first.

**Comment by nmbradford** [comment]
Lots or parameters aren't a problem, as long as the code is clean and the invocation is not confusing for users.  I don't see anything here that is an issue.  Also - given the comment inline about the fact that you can use the full edge api to update clisettings, i think it makes more sense in set-nsxedge.

**Comment by alagoutte** [comment]
Move all set-nsxclisettings to set-nsxedge and remove get-nsxclisettings (or always kept ?)

**Comment by nmbradford** [comment]
if you think its needed, im not opposed to having get-nsx_edge_clisettings, but given you can just go (get-nsxedge fred).cliSettings, I'm not sure I see the value! :D

**Comment by alagoutte** [comment]
Ok, i will update my patches soon...

Can you look some other (small) PR (like #454 #455 #466)

**Comment by alagoutte** [comment]
Updated (test suite and some other change)

it is better ?

**Comment by nmbradford** [comment]
password is still mandatory - it needs to not be, I showed you how to avoid that requirement in the API, and it is still a separate cmdlet for set - it needs to be part of set-nsxedge as previously commented.

**Comment by alagoutte** [comment]
i ask about test suite if it is good...

**Comment by alagoutte** [comment]
Hi @nmbradford, can you look ? 

it is add now on set-nsxedge (but kept set-clisettings...)

**Comment by alagoutte** [comment]
@nmbradford any news ?

**Comment by nmbradford** [comment]
looks a lot better (havent had a chance to review the tests yet, but the changes in Set-NsxEdge look good.

What is still a problem is that you have (nearly) duplicate code in Set-CliSettings (which is still also called the wrong thing as it doesnt include the word edge in it.) and _still_ has a mandatory password.

Get-CliSettings is also called the wrong thing.

You need to either 
a) Remove Get/Set-CliSettings (what value does it serve over Get/Set-NsxEdge now?)
b) Rename and modify Get/Set-CliSettings 👍 (include the word edge).  Remove duplicate code (call set-NsxEdge internally rather than doing the same thing twice.  Make password non-mandatory.

**Comment by alagoutte** [comment]
i will look to avoid duplicate code

I prefer to kept name Get/set-nsxCliSettings because i can be also working for DLR.. (and not also Edge)
and for other Edge Feature, there is not Edge on cmdlets..

**Comment by nmbradford** [comment]
There is a CLI in NSX other than edge/dlr so this wont be ok.  If you are adamant about keeping a cli specific cmdlet, please name it as requested.  Create a second set of cmdlets that are DLR specific that also wraps the same cmd.  As repeatedly stated, my preference is for there to be no cli specific cmdlets.  This is yet another reason for that.

**Comment by alagoutte** [comment]
Remove [Get|Set]-nsxCliSettings and only kept Set-NsxEdge with cliSettings

**Comment by alagoutte** [comment]
@dcoghlan can you launch test ? (and merge if it is good ?)

**Comment by dcoghlan** [comment]
jenkins test this please

File: module/PowerNSX.psm1 [hunk 16]
```diff
@@ -13916,6 +13993,51 @@ function Set-NsxEdge {
...
+            } else {
+                Add-XmlElement -xmlroot $_Edge.cliSettings -xmlElementName "remoteAccess" -xmlElementText $remoteAccess.ToString().ToLower()
+            }
+        }
+
+        if ( $PsBoundParameters.ContainsKey('passwordExpiry') ) {
+            if ( invoke-xpathquery -node $_Edge -querymethod SelectSingleNode -Query "child::cliSettings/passwordExpiry" ) {
+                $_Edge.cliSettings.passwordExpiry = $passwordExpiry
+            } else {
+                Add-XmlElement -xmlroot $_Edge.cliSettings -xmlElementName "passwordExpiry" -xmlElementText $passwordExpiry
```
Comment (dcoghlan) [inline module/PowerNSX.psm1]: This line fails the windows tests
```
      [-] Change Password Expiry 277ms
        Cannot set "passwordExpiry" because only strings can be used as values to set XmlNode properties.
        at <ScriptBlock>, C:\jenkins\workspace\WindowsPowerNsx636\tests\integration\04.Edge.Tests.ps1: line 762
        762:             Get-NsxEdge $name | Set-NsxEdge -passwordExpiry 4242 -confirm:$false
```
Change to `$passwordExpiry.ToString()` and it should pass the windows tests.

**Comment by dcoghlan** [comment]
jenkins test this please

**Comment by alagoutte** [comment]
it is ok ?

**Comment by dcoghlan** [comment]
jenkins test this please

File: module/PowerNSX.psm1 [hunk 17]
```diff
@@ -13947,6 +14024,51 @@ function Set-NsxEdge {
...
+            if ( invoke-xpathquery -node $_Edge -querymethod SelectSingleNode -Query "child::cliSettings/remoteAccess" ) {
+                $_Edge.cliSettings.remoteAccess = $remoteAccess.ToString().ToLower()
+            } else {
+                Add-XmlElement -xmlroot $_Edge.cliSettings -xmlElementName "remoteAccess" -xmlElementText $remoteAccess.ToString().ToLower()
+            }
+        }
+
+        if ( $PsBoundParameters.ContainsKey('passwordExpiry') ) {
+            if ( invoke-xpathquery -node $_Edge -querymethod SelectSingleNode -Query "child::cliSettings/passwordExpiry" ) {
+                $_Edge.cliSettings.passwordExpiry = $passwordExpiry
```
Comment (dcoghlan) [inline module/PowerNSX.psm1]: This also needs to be cast as a string.

Comment (alagoutte) [reply module/PowerNSX.psm1]: Done

Comment (alagoutte) [reply module/PowerNSX.psm1]: On what release it don't work ? (because work for me on Windows with PowerShell 6/Core

**Comment by dcoghlan** [comment]
jenkins test this please

File: module/PowerNSX.psm1 (hunk 17 above)
Comment (dcoghlan) [reply module/PowerNSX.psm1]: needs to work on Powershell 5.1 on Windows

**Comment by dcoghlan** [comment]
jenkins test this please

**Comment by dcoghlan** [comment]
CLI tests have passed now. Overall tests failed due to cleanup error, so this one is good to go now.
